Lesson_19/homework_19_1.py

```diff
@@ -0,0 +1,145 @@
+# photos taken by the Curiosity rover on Mars based on certain parameters.
+# Among this data are links to photos that need to be parsed and then,
+# using additional queries, download and save
+# these photos as local files mars_photo1.jpg, mars_photo2.jpg.
+# The task needs to be done using the requests module
+import json
+import logging
+
+import requests
+
+
+def configure_logging():
+    """Logger configuration.
+
+    Returns:
+        logging.Logger: Logging Logger.
+    """
+    logger = logging.getLogger(__name__)
+    logger.setLevel(logging.DEBUG)
+
+    console_handler = logging.StreamHandler()
+    console_handler.setLevel(logging.DEBUG)
+
+    log_formatter = logging.Formatter(
+        ('%(asctime)s.%(msecs)03d : %(module)-12s: '
+         '%(lineno)3d %(levelname)-5s - %(message)s'),
+        '%Y-%m-%d %H:%M:%S',
+    )
+
+    console_handler.setFormatter(log_formatter)
+    logger.addHandler(console_handler)
+
+    return logger
+
+
+_log = configure_logging()
+
+
+class MarsRoverPhotos:
+    """A class to interact with NASA's Mars Rover Photos API.
+
+    Attributes:
+        url (str): The API endpoint URL.
+        params (dict): Parameters to be sent with the API request.
+        timeout (int): Timeout duration for HTTP requests in seconds.
+    """
+
+    def __init__(self, url, params):
+        """Initialize the MarsRoverPhotos instance.
+
+        Args:
+            url (str): The API endpoint URL.
+            params (dict): Parameters to be sent with the API request.
+        """
+        self.url = url
+        self.params = params
+        self.timeout = 5
+
+    def fetch_photos(self):
+        """Fetch photos metadata from NASA API."""
+        try:
+            # Send request to NASA API to get photos data
+            response = requests.get(
+                self.url, self.params, timeout=self.timeout)
+            response.raise_for_status()  # Rase an exception if code is not 2xx
+
+            # Get JSON data from response
+            try:
+                data = response.json()
+                photos = data.get('photos')
+
+                if photos:
+                    self.download_photos(photos)
+                else:
+                    _log.info('No photos found.')
+
+            except json.JSONDecodeError as error:
+                error = 'Error serializing JSON: {error}'
+                _log.error(error)
+        except requests.exceptions.HTTPError as error:
+            error = f'HTTP Error: {error}'
+            _log.error(error)
+        except requests.exceptions.ConnectionError as error:
+            error = f'Connection Error: {error}'
+            _log.error(error)
+        except requests.exceptions.Timeout as error:
+            error = f'Timeout Error: {error}'
+            _log.error(error)
+        except requests.exceptions.RequestException as error:
+            error = f'Request failed: {error}'
+            _log.error(error)
```

Comment:
You have the same errors handling in second call - better to organize single function aka do_request and use for both methods.

Reply:
Fixed in commit 31dfb0b758979996cd1a5d4f706a0d589d4b24f9